subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java

```diff
@@ -128,16 +129,19 @@ private static boolean missesDependency(Node producer, Node consumer) {
         }
         // Do a breadth first search for any dependency
         Deque<Node> queue = new ArrayDeque<>();
-        consumer.getHardSuccessors().forEach(queue::add);
+        Set<Node> seenNodes = new HashSet<>();
+        consumer.getHardSuccessors().forEach(successor -> {
+            if (seenNodes.add(successor)) {
+                queue.add(successor);
+            }
+        });
```

Comment:
I think this is fine as it is, though it might be worth extracting this logic somehow. What would best capture the intention here is if we wrapped the queue into something like a `ProcessOnceQueue` that hid the deduplication completely.

Reply:
If you have some other places in mind where we use such a queue, then I would extract the logic. If this is the only place, I wouldn't. So if we stumble on another occasion where we need such a queue, we should extract it.

Reply:
I think extracting the logic by itself is already useful just for being able to name the concept. That said, we use the same logic in `Types.walkTypeHierarchy()` and in `ModelTypes.collectHierarchy()`.

In most places we use `xxxSeen` and not `seenXxx`.

Reply:
The reason I think it would be useful to extract here is because unlike the other places here we add to the queue in two different locations, and that is more error-prone.

Reply:
I'll prepare a separate PR where I extract the queue.